[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos, add support for param dedup

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95737

This is an iteration on a prior implementation that used positions. This is more sound, and easier to follow, but does have an open question w/r/t flattening.

A few thoughts / discussion points / notes:

1) Writing source to a tensor is not good enough. Tensor identity is the very nature by which we dedup, so all duplicate tensors ARE the same tensor - writing a source to one means the last source wins. There is a potential alternative impl that instead replace writing source with writing a list of sources, but that kind of puts the notion of duplication into that list. 

2) Placeholders alone is not good enough. We need params and buffers too, and the param and buffer lifting happens downstream of dynamo. 

3) This feels a little bit brittle because it still relies on position matching, but with a much simpler contract since the positions are not used for keying or indexing, and instead just rely on a match between params+buffers+inputs and the gm.params+gm.buffers+example args, which feels like a much tighter contract.

4) I do not know enough about AOT calling conventions to understand the flattening gaps here - ostensibly, dynamo should never be sending inputs to aot_autograd where flattening matters, but I do not know. cc @ezyang 

`pytest test/dynamo/test_aot_autograd.py -k test_arg_dupe_via_dynamo_recompiles_many_args`

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos

Lint

[ghstack-poisoned]

wconstab commented on torch/_dynamo/guards.py:
```diff
@@ -643,19 +643,9 @@ def compile_check_fn(
-                code_part = f"{self.output_graph.graphargs[pos_a].source.name()} is {self.output_graph.graphargs[pos_b].source.name()}"  # noqa: B950
+                source_a = guard.input_source_a
+                source_b = guard.input_source_b
+                code_part = f"{source_a.name()} is {source_b.name()}"
```
i feel like it'd be good to be explicit about matching the logic used here (a is b) with the logic used in aot when generating the dupe args map in the first place.

a way to do that nicely might be to factor it out as a function `is_duplicate(a, b)` which we call in both places.

a related question is what is the correct implementation for `is_duplicate` - should it care about pyobj or c++ tensorImpl or some alias property or what?

(i think we can probably land this as is first and follow up, since it seems strictly better than status quo)

wconstab reviewed (approved):
lgtm, thanks @voznesenskym !

voznesenskym committed:
Update on "[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos"


This is an iteration on a prior implementation that used positions. This is more sound, and easier to follow, but does have an open question w/r/t flattening.

A few thoughts / discussion points / notes:

1) Writing source to a tensor is not good enough. Tensor identity is the very nature by which we dedup, so all duplicate tensors ARE the same tensor - writing a source to one means the last source wins. There is a potential alternative impl that instead replace writing source with writing a list of sources, but that kind of puts the notion of duplication into that list. 

2) Placeholders alone is not good enough. We need params and buffers too, and the param and buffer lifting happens downstream of dynamo. 

3) This feels a little bit brittle because it still relies on position matching, but with a much simpler contract since the positions are not used for keying or indexing, and instead just rely on a match between params+buffers+inputs and the gm.params+gm.buffers+example args, which feels like a much tighter contract.

4) I do not know enough about AOT calling conventions to understand the flattening gaps here - ostensibly, dynamo should never be sending inputs to aot_autograd where flattening matters, but I do not know. cc ezyang 

`pytest test/dynamo/test_aot_autograd.py -k test_arg_dupe_via_dynamo_recompiles_many_args`

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos"


This is an iteration on a prior implementation that used positions. This is more sound, and easier to follow, but does have an open question w/r/t flattening.

A few thoughts / discussion points / notes:

1) Writing source to a tensor is not good enough. Tensor identity is the very nature by which we dedup, so all duplicate tensors ARE the same tensor - writing a source to one means the last source wins. There is a potential alternative impl that instead replace writing source with writing a list of sources, but that kind of puts the notion of duplication into that list. 

2) Placeholders alone is not good enough. We need params and buffers too, and the param and buffer lifting happens downstream of dynamo. 

3) This feels a little bit brittle because it still relies on position matching, but with a much simpler contract since the positions are not used for keying or indexing, and instead just rely on a match between params+buffers+inputs and the gm.params+gm.buffers+example args, which feels like a much tighter contract.

4) I do not know enough about AOT calling conventions to understand the flattening gaps here - ostensibly, dynamo should never be sending inputs to aot_autograd where flattening matters, but I do not know. cc ezyang 

`pytest test/dynamo/test_aot_autograd.py -k test_arg_dupe_via_dynamo_recompiles_many_args`

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
I kinda hate this PR now >:(

voznesenskym committed:
Update on "[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos"


This is an iteration on a prior implementation that used positions. This is more sound, and easier to follow, but does have an open question w/r/t flattening.

A few thoughts / discussion points / notes:

1) Writing source to a tensor is not good enough. Tensor identity is the very nature by which we dedup, so all duplicate tensors ARE the same tensor - writing a source to one means the last source wins. There is a potential alternative impl that instead replace writing source with writing a list of sources, but that kind of puts the notion of duplication into that list. 

2) Placeholders alone is not good enough. We need params and buffers too, and the param and buffer lifting happens downstream of dynamo. 

3) This feels a little bit brittle because it still relies on position matching, but with a much simpler contract since the positions are not used for keying or indexing, and instead just rely on a match between params+buffers+inputs and the gm.params+gm.buffers+example args, which feels like a much tighter contract.

4) I do not know enough about AOT calling conventions to understand the flattening gaps here - ostensibly, dynamo should never be sending inputs to aot_autograd where flattening matters, but I do not know. cc ezyang 

`pytest test/dynamo/test_aot_autograd.py -k test_arg_dupe_via_dynamo_recompiles_many_args`

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
> I kinda hate this PR now >:(

Maybe we just make the logical equivalent of `graph_arg_pos` on an flattened tensor? And build the dupe list as we go in dynamo? I dunno....

voznesenskym committed:
Update on "[WIP][Draft][Discussion] AOT Autograd dedup by source, not pos, add support for param dedup"


This is an iteration on a prior implementation that used positions. This is more sound, and easier to follow, but does have an open question w/r/t flattening.

A few thoughts / discussion points / notes:

1) Writing source to a tensor is not good enough. Tensor identity is the very nature by which we dedup, so all duplicate tensors ARE the same tensor - writing a source to one means the last source wins. There is a potential alternative impl that instead replace writing source with writing a list of sources, but that kind of puts the notion of duplication into that list. 

2) Placeholders alone is not good enough. We need params and buffers too, and the param and buffer lifting happens downstream of dynamo. 

3) This feels a little bit brittle because it still relies on position matching, but with a much simpler contract since the positions are not used for keying or indexing, and instead just rely on a match between params+buffers+inputs and the gm.params+gm.buffers+example args, which feels like a much tighter contract.

4) I do not know enough about AOT calling conventions to understand the flattening gaps here - ostensibly, dynamo should never be sending inputs to aot_autograd where flattening matters, but I do not know. cc ezyang 

`pytest test/dynamo/test_aot_autograd.py -k test_arg_dupe_via_dynamo_recompiles_many_args`

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
from a brief skim: let's do the source plumbing on its own, and describe how the design for that is supposed to work (with reference to the dedup problem, but let's make sure the design stands on its own)
